clients/src/main/java/org/apache/kafka/common/record/KafkaLZ4BlockInputStream.java

```diff
@@ -145,46 +154,72 @@ private void readHeader() throws IOException {
      * @throws IOException
      */
     private void readBlock() throws IOException {
-        int blockSize = ByteUtils.readUnsignedIntLE(in);
+        if (in.remaining() < 4) {
+            throw new IOException(PREMATURE_EOS);
+        }
+
+        int blockSize = in.getInt();
+        boolean compressed = (blockSize & LZ4_FRAME_INCOMPRESSIBLE_MASK) == 0;
+        blockSize &= ~LZ4_FRAME_INCOMPRESSIBLE_MASK;
```

Comment:
Indeed, the size check wasn't working, but at the end of the day, we would just thrown a different exception in case the block size exceeded the max.

Regarding the zero size block, we were just lucky that we never set the `LZ4_FRAME_INCOMPRESSIBLE_MASK` flag in `KafkaLZ4BlockOutputStream.writeEndMark`. I checked the spec and it didn't seem to specify whether the flag should be set or not in that case, so better be safe.